src/models.jl

```diff
@@ -0,0 +1,124 @@
+mutable struct Model
+    metadata::ModelMetadata
+    configuration::ModelConfiguration
+    boundary_conditions::BoundaryConditions
+    constants::PlanetaryConstants
+    eos::EquationOfState
+    grid::Grid
+    velocities::VelocityFields
+    tracers::TracerFields
+    pressures::PressureFields
+    G::SourceTerms
+    Gp::SourceTerms
+    forcings::ForcingFields
+    forcing::Forcing
+    stepper_tmp::StepperTemporaryFields
+    operator_tmp::OperatorTemporaryFields
+    ssp  # ::SpectralSolverParameters or ::SpectralSolverParametersGPU
+    clock::Clock
+    output_writers::Array{OutputWriter,1}
+    diagnostics::Array{Diagnostic,1}
+end
+
+"""
+    Model(; kwargs...)
+
+Construct an `Oceananigans.jl` model. The keyword arguments are:
+
+          N (tuple) : model resolution in (x, y, z)
+          L (tuple) : model domain in (x, y, z)
+         dt (float) : time step
+ start_time (float) : start time for the simulation
+  time_step (float) : ?
+         arch (sym) : architecture (:cpu or :gpu)
+  float_type (type) : floating point type for model data (typically Float32 or Float64)
+          constants : planetary constants (?)
+                eos : equation of state to infer density from temperature and salinity
+            forcing : forcing functions for (u, v, w, T, S)
+boundary_conditions : boundary conditions
+     output_writers : output writer
+        diagonstics : diagnostics
+
+        ... and more.
+"""
+function Model(;
+    # Model resolution and domain size
+             N = (32, 1, 32),
+             L = (1.0, 1.0, 1.0),
+    # Molecular parameters
+             ν = 1.05e-6,
+            νh = ν,
+            νv = ν,
+             κ = 1.43e-7,
+            κh = κ,
+            κv = κ,
+    # Time stepping
+            dt = 1e-16,
+    start_time = 0,
+     time_step = 0, # ?
+    # Model architecture and floating point precision
+          arch = :cpu,
+    float_type = Float64,
+     constants = Earth(),
+    # Equation of State
+           eos = LinearEquationOfState(),
+    # Forcing and boundary conditions for (u, v, w, T, S)
+       forcing = Forcing(nothing, nothing, nothing, nothing, nothing),
+    boundary_conditions = BoundaryConditions(:periodic, :periodic, :rigid_lid, :free_slip),
+    # Output and diagonstics
+    output_writers = OutputWriter[],
+       diagnostics = Diagnostic[]
+)
+
+    # Initialize model basics.
+         metadata = ModelMetadata(arch, float_type)
+    configuration = ModelConfiguration(νh, νv, κh, κv)
+             grid = RegularCartesianGrid(metadata, N, L)
+            clock = Clock(start_time, time_step, dt)
```

Comment:
I think the user should also specify `Δt` at a minimum. Having `Δt = 1e-16` by default would produce confusing behavior in my opinion, e.g. why is my model not doing anything?

Reply:
I actually was not aware that you could specify keyword arguments without defaults.

However, I also think we should remove `dt` from `Clock`. Then we would not have to specify the time-step in the model constructor.

Reply:
Yes that makes more sense. `Clock` should just keep track of time and iteration number, but not the time step.